### PR TITLE
Client stability improvements

### PR DESCRIPTION
**Pull request by khayalan-mathew**
This PR aims to make the remote & local clients more stable.
Especially the remote client has received some security measurements:
It now follows a pattern, where the person/edge-engine spinning up the tec engine is the owner.
Only the owner can stop the engine and decide on the settings, but regardless other product engines can connect and use the output.

Due to the remote client not used in production, these limits are intended to provide a better development process and not fail-safety.

Additionally, the library now generates source maps to make debugging easier and the connect/disconnect methods now return a Promise so you can detect when it actually disconnected properly.

### CODE REVIEW

File: src/io/IO.ts [hunk 1]
```diff
@@ -103,19 +103,35 @@ export class IO {
    * Starts monitoring the messages
    * @param {IOOptions} options
    */
-  public connect(options: IOOptions = {}): void {
-    Logger.log(`Client initializing...`);
-    this.client.init();
-    if (!options.noSnapshot) {
-      this.poiMonitor.start();
-    }
+  public async connect(options: IOOptions = {}): Promise<void> {
```
Comment (micaelmbagira) [inline src/io/IO.ts]: Not async 😀

File: src/io/IO.ts [hunk 2]
```diff
@@ -103,19 +103,35 @@ export class IO {
...
+      this.client.init();
+      if (!options.noSnapshot) {
+        this.poiMonitor.start();
+      }
+      const resolveCb = () => {
+        this.client.removeListener(ClientEvent.Init, resolveCb);
+        resolve();
+      };
+      this.client.on(ClientEvent.Init, resolveCb);
+      this.client.close();
```
Comment (micaelmbagira) [inline src/io/IO.ts]: Why closing the client?

Comment (micaelmbagira) [inline src/io/IO.ts]: Why closing the client?

Comment (khayalan-mathew) [reply src/io/IO.ts]: In case we want to disconnect the IO we want to disconnect the connection of the client as well, do we?

Comment (micaelmbagira) [reply src/io/IO.ts]: This is the `connect` method

Comment (khayalan-mathew) [reply src/io/IO.ts]: Right stupid mistake.
-> Fixed

File: src/client/RemoteClient.ts [hunk 3]
```diff
@@ -131,6 +141,8 @@ export class AdvertimaRemoteClient extends AdvertimaClient {
   private onInit(message?: any) {
     if (message['result'] === 'success') {
       this.ipcInitialized = true;
+    } else {
+      this.close;
```
Comment (micaelmbagira) [inline src/client/RemoteClient.ts]: Typo, should be `this.close()`

File: src/client/LocalClient.ts [hunk 4]
```diff
@@ -140,12 +142,13 @@ export class AdvertimaLocalClient extends AdvertimaClient {
...
     this.process = null;
-    this.emit(ClientEvent.Close, {
-      expectedShutdown: !this.shouldBeRunning,
-    });
+    const expectedShutdown = !this.shouldBeRunning;
     if (this.shouldBeRunning) {
       this.init();
     }
+    this.emit(ClientEvent.Close, {
+      expectedShutdown,
```
Comment (micaelmbagira) [inline src/client/LocalClient.ts]: Do we only need this `expectedShutdown` for the stress testing?

File: src/client/RemoteClient.ts [hunk 5]
```diff
@@ -214,7 +232,13 @@ export class AdvertimaRemoteClient extends AdvertimaClient {
...
-    this.ipc.disconnect(this.clientToken);
-    this.ipcInitialized = false;
+    this.shouldBeRunning = false;
+    this.rpc(
+      RPCFunction.Close,
+      {},
+      {
+        noResponseExpected: true,
+      }
+    );
```
Comment (micaelmbagira) [inline src/client/RemoteClient.ts]: Maybe we can have this statement in one line `this.rpc(RPCFunction.Close, {}, { noResponseExpected: true });`

File: src/client/LocalClient.ts (hunk 4 above)
Comment (khayalan-mathew) [reply src/client/LocalClient.ts]: The close event gets emitted every time the client closes regardless of if it is a crash or a shutdown.
This message is to differentiate those two.

But yes currently it is just used for stress testing.

**Pull request merged**